java-generator/core/src/main/java/io/fabric8/java/generator/FileJavaGenerator.java

```diff
@@ -83,32 +85,50 @@ private void runOnSingleSource(File source, File basePath) {
       }
 
-      resources.parallelStream()
-          .forEach(
-              rawResource -> {
-                if (rawResource != null && rawResource instanceof HasMetadata) {
-                  final HasMetadata resource = (HasMetadata) rawResource;
+      writeCRCompilationUnits(basePath, generateWritableCRCompilationUnits(resources));
 
-                  if (resource != null && resource.getKind()
-                      .toLowerCase(Locale.ROOT)
-                      .equals("customresourcedefinition")) {
-                    CustomResourceDefinition crd = (CustomResourceDefinition) resource;
-
-                    final String basePackage = groupToPackage(crd.getSpec().getGroup());
-
-                    crGeneratorRunner.generate(crd, basePackage).parallelStream()
-                        .forEach(w -> w.writeAllJavaClasses(basePath));
-                  } else {
-                    LOGGER.warn("Not generating nothing for resource of kind: {}", resource.getKind());
-                  }
-                } else {
-                  LOGGER.warn("Not generating nothing for unrecognized resource: {}", Serialization.asYaml(rawResource));
-                }
-              });
     } catch (FileNotFoundException e) {
       throw new JavaGeneratorException("File " + source.getAbsolutePath() + " not found", e);
     } catch (IOException e) {
       throw new JavaGeneratorException("Exception reading " + source.getAbsolutePath(), e);
     }
   }
+
+  private List<WritableCRCompilationUnit> generateWritableCRCompilationUnits(List<Object> resources) {
+    return resources.parallelStream()
+        .flatMap(
+            rawResource -> {
+              if (rawResource != null && rawResource instanceof HasMetadata) {
+                final HasMetadata resource = (HasMetadata) rawResource;
+
+                if (resource != null && resource.getKind()
+                    .toLowerCase(Locale.ROOT)
+                    .equals("customresourcedefinition")) {
+                  CustomResourceDefinition crd = (CustomResourceDefinition) resource;
+
+                  final String basePackage = groupToPackage(crd.getSpec().getGroup());
+                  return crGeneratorRunner.generate(crd, basePackage).stream();
+                } else {
+                  LOGGER.warn("Not generating nothing for resource of kind: {}", resource.getKind());
+                }
+              } else {
+                LOGGER.warn("Not generating nothing for unrecognized resource: {}", Serialization.asYaml(rawResource));
+              }
+              return Stream.empty();
+            })
+      .collect(Collectors.toList());
+  }
+
+  private void writeCRCompilationUnits(File basePath, List<WritableCRCompilationUnit> wCUs) {
+    CompletableFuture<Void>[] futures = new CompletableFuture[wCUs.size()];
+
+    int i = 0;
+    for (WritableCRCompilationUnit w : wCUs) {
+      futures[i++] = CompletableFuture.runAsync(() -> w.writeAllJavaClasses(basePath));
```

Comment:
Have you attempted to use a different Thread Pool than the `ForkJoinPool.commonPool` since it's used also by the underlying `parallelStreams`?

Reply:
At that point the code generation performed in the parallel stream is finished and that pool should be completely available to write files. Also consider that creating a brand new thread pool could be very likely much more expansive than writing the file, that's why imo it is a better idea to reuse an already existing pool. Anyway if we had the JMH benchmark that I'm proposing in the other comment we could properly measure this.

Reply:
You can find the yaml I was using here: https://github.com/kserve/kserve/releases/download/v0.10.0/kserve.yaml
feel free to add it to the `CompilationTest` along with this PR.

Feel free to add a JMH benchmark if you think that it's useful, but I'm convinced that this PR is perfectly enough for a first iteration.
In case performance issues becomes a problem we can revisit the implementation later on :+1:

Thanks again!